### ISSUE TITLE
Gadgetron configuration not validated properly

### ISSUE THREAD

**Issue by naegelejd** [gadgetron/gadgetron#185]
The recent CloudBus changes changed the name of the XML element for specifying the `relayAddress` (formerly `multiCastAddress`. The schema appears to be correct, however if I start the gadgetron with an old configuration file (containing `multiCastAddress`), there are no complaints and I get a bunch of ACE_INET_Addr errors because it's trying to use an "empty" address.

I haven't looked very hard but it appears the configuration file isn't being validated against the schema.

**Comment by hansenms** [comment]
We no longer validate this since we stopped using xsd stuff to parse the XML. 

I think this is a bug in the sense that when there is no relayAddress, it should not attempt to connect to connect. Let me have a look.

**Comment by hansenms** [comment]
Actually, looking at it now, it defaults to "localhost". Since localhost is a bit non-specific (and fails on many Linux distros), we try to get the name of the host with this code:

    ACE_SOCK_Acceptor listener (ACE_Addr::sap_any);
    ACE_INET_Addr local_addr;
    listener.get_local_addr (local_addr);
    node_info_.address = std::string(local_addr.get_host_name());

On out NIH macs this is your machine name (some strange name given to you by the DNS) and if you are not on the NIHnetwork, your mac will not resolve this to an IP and that is what you are getting errors from.  If you make sure that your host name is in your /etc/hosts file or something you will not get the error. 

Alternatively, you should make a proper gadgetron.xml file and put 127.0.0.1 as the relay address. 

So the error is not cause by a failure to validate the document, it is caused by your laptop not being able to resolve it's own hostname to an IP address. 

Makes sense?

**Comment by naegelejd** [comment]
I'm running the Gadgetron on an iMac that is always on the NIH network:
```
$ MH01684470MACDT ~ hostname
MH01684470MACDT
$ MH01684470MACDT ~ nslookup `hostname`
Server:		156.40.70.10
Address:	156.40.70.10#53

Name:	MH01684470MACDT.nih.gov
Address: 156.40.186.134

$ MH01684470MACDT ~ ping `hostname`
PING mh01684470macdt.nih.gov (156.40.186.134): 56 data bytes
64 bytes from 156.40.186.134: icmp_seq=0 ttl=64 time=0.042 ms
64 bytes from 156.40.186.134: icmp_seq=1 ttl=64 time=0.079 ms
64 bytes from 156.40.186.134: icmp_seq=2 ttl=64 time=0.069 ms
64 bytes from 156.40.186.134: icmp_seq=3 ttl=64 time=0.062 ms
^C
--- mh01684470macdt.nih.gov ping statistics ---
4 packets transmitted, 4 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 0.042/0.063/0.079/0.014 ms
$ MH01684470MACDT ~ gadgetron
2015-07-14 16:59:38 INFO [main.cpp:161] Starting cloudBus: :8002
2015-07-14 16:59:38 INFO [main.cpp:188] Configuring services, Running on port 9002
ACE_INET_Addr::ACE_INET_Addr: : Undefined error: 0
ACE_INET_Addr::ACE_INET_Addr: : Can't assign requested address
ACE_INET_Addr::ACE_INET_Addr: : Can't assign requested address
ACE_INET_Addr::ACE_INET_Addr: : Can't assign requested address
^C
$ MH01684470MACDT ~ 
```

**Comment by hansenms** [comment]
Try to ping MH01684470MACDT

That is what your host name looks like and I am pretty sure it does not resolve.

------

**Comment by naegelejd** [comment]
I just pasted output of ping above... it's the same command:
```
$ MH01684470MACDT ~ ping MH01684470MACDT
PING mh01684470macdt.nih.gov (156.40.186.134): 56 data bytes
64 bytes from 156.40.186.134: icmp_seq=0 ttl=64 time=0.041 ms
64 bytes from 156.40.186.134: icmp_seq=1 ttl=64 time=0.097 ms
64 bytes from 156.40.186.134: icmp_seq=2 ttl=64 time=0.063 ms
64 bytes from 156.40.186.134: icmp_seq=3 ttl=64 time=0.060 ms
^C
--- mh01684470macdt.nih.gov ping statistics ---
4 packets transmitted, 4 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 0.041/0.065/0.097/0.020 ms
$ MH01684470MACDT ~ 
```
No issues resolving whatsoever.

**Comment by hansenms** [comment]
Strange. Not really sure how to debug that. I'll try to see what I can do.

------

**Comment by hansenms** [comment]
I can't actually reproduce this problem right now on my machine. It works just fine regardless.

**Comment by naegelejd** [comment]
Interesting. Let's close it then.

**Comment by hansenms** [comment]
Well, I have seen it before, so I would kinda like to take a look at it. 

Can you just confirm for me when you have a relay address set in your gadgetron.xml you don't see it?

Can you also tell me if you put "localhost" in your relay address, do you see it?

Also if you have a situation where you see it, could you edit toolboxes/cloudbus/CloudBus.cpp:

    listener.get_local_addr (local_addr);
    node_info_.address = std::string(local_addr.get_host_name());
    node_info_.active_reconstructions = 0;
 
to

    listener.get_local_addr (local_addr);
    node_info_.address = std::string(local_addr.get_host_name());
    GDEBUG("node_info_.address : %s\n", node_info_.address.c_str())
    node_info_.active_reconstructions = 0;

And tell me what the debug output says your hostname is. Can you ping that hostname?

**Comment by naegelejd** [comment]
If I set a relay address of 127.0.0.1 or "localhost" I have no problems. If I set an empty relay address, or omit `<relayAddress>` altogether, I see the error.

I think this behavior makes sense, but a more useful error message for an invalid relay address could be helpful.

**Comment by hansenms** [comment]
What if you omit the cloudbus section altogether?

There may be a case we are not handling, i.e. No address but cloudbus is there. That should be a parse error, I can fix that.

No cloudbus section should cause it to not connect.

------

**Comment by naegelejd** [comment]
Yeah if the cloudbus section is removed there aren't any cloudbus connection attempts.

**Comment by hansenms** [comment]
Ok, I'll make a fix later.

------

**Comment by hansenms** [comment]
Here is the fix. 

https://github.com/gadgetron/gadgetron/pull/186

Please close when merged.

**Comment by naegelejd** [comment]
fixed by #186
